### ISSUE TITLE
GenericBioEngineApp: more robust model prediction

### ISSUE THREAD

**Issue by constantinpape** [bioimage-io/archive.bioimage.io#290]
I noticed that we are using the prediction pipeline directly in model prediction on the server. This means that the input shape needs to fit the expected model shape exactly. We should instead check that the shapes match and if not use padding or tiling to enable inference.
All of this functionality is implemented in bioimageio.core already, I will provide links later.

**Comment by constantinpape** [comment]
I just made a new release in bioimageio.core that makes all of this easier to use. The following should work now:
```py
if bioimageio.core.check_input_shape(input.shape, model.inputs[0].spec.shape):
  pred = ...   # shape is valid, we can just run prediction as before
else:
  pred = bioimageio.core.predict_with_padding(pred_pipeline, input)  # predict with padding
  # or
  pred = bioimageio.core.predict_with_tiling(pred_pipeline, input)
```

**Comment by oeway** [comment]
This is great! I will integrate this to the server. To make sure we don't get OOM error, maybe we should always do tiling on the server? Otherwise, what's the strategy to decide when to do tiling or padding?

**Comment by constantinpape** [comment]
> This is great! I will integrate this to the server. To make sure we don't get OOM error, maybe we should always do tiling on the server? Otherwise, what's the strategy to decide when to do tiling or padding?

Yea, I think going with tiling if the shape doesn't fit is a good idea!

**Comment by constantinpape** [comment]
Hmm, thinking about this: for models with a implicit shape descriptions we can of course also get valid shapes that would result in an OOM.
Maybe we should add some heuristics to determine whether or not to do tiling in general and then do something like this:
```py
use_tiling = tiling_heuristics(input.shape)
valid_shape = bioimageio.check_input_shape(input.shape, model.inputs[0].spec.shape
if use_tiling:
  pred = ... # predict with tiling
elif valid_shape:
  pred = ... # normal prediction
else:
  pred = ... # predict with padding  
```
And as heuristic we could use something fairly simple, e.g. `np.prod(input.shape) <= 32 * 256 * 256` (that's the size of a 3d block that usually fits a large 3d unet and I tried that successfully for some 3d models already).

**Comment by oeway** [comment]
Can we do tiling and padding on the client side? Two reasons:

1. It's more stable to send data by small tiles, and we can scale to much larger data size. E.g later we can easily support NGFF or large tissue image.

2. In the next version of bioengine, we will convert all the models in the triton supported format, and thus allows dynamic batching, meaning the server will try to process data in batches even if requests are from different users.

**Comment by oeway** [comment]
Hi again, I saw you did: `if bioimageio.core.check_input_shape(input.shape, model_resource.inputs[0].spec.shape)`, how do we handle the case when we have multiple inputs? Is there a more generic version for this, check for all the inputs and do `any(...)`?

**Comment by constantinpape** [comment]
> Can we do tiling and padding on the client side? Two reasons:
> 
>     1. It's more stable to send data by small tiles, and we can scale to much larger data size. E.g later we can easily support NGFF or large tissue image.
> 
>     2. In the next version of bioengine, we will convert all the models in the triton supported format, and thus allows dynamic batching, meaning the server will try to process data in batches even if requests are from different users.

I fully agree with this in principle. But I think the best way to do this would be a `TritonModelAdapter` in bioimageio.core. And that's probably a bit more work (and we don't even have the triton models yet). So I think we should first make the changes I proposed above; and then switch to your suggestion later.


> Hi again, I saw you did: `if bioimageio.core.check_input_shape(input.shape, model_resource.inputs[0].spec.shape)`, how do we handle the case when we have multiple inputs? Is there a more generic version for this, check for all the inputs and do `any(...)`?

yeah, that's easy (but needs to be `all` instead of `any`):
```py
have_valid_shape = all(bioimageio.core.check_input_shape(input.shape, input_spec.shape) for input, input_spec in zip(inputs, model_resource.inputs))
```

**Comment by constantinpape** [comment]
@oeway fyi, new version is available on conda-forge now
